### PR TITLE
feat: add union types

### LINKED ISSUE #24: Add support for union types
A union type is a set consisting of multiple non-overlapping types. With union types, we can have function signatures of the type:

```elixir
fn foo : {:ok, Int} | {:error, String} do
  if true do
    {:ok, 123}
  else
    {:error, "Something went wrong"}
  end
end
```

We'll need to understand the concept of subtypes before we start working on this. [This wiki](https://en.wikipedia.org/wiki/Subtyping) is a good introduction.

In this example:
image
duck is a subtype of bird (which is its supertype), which means it can be substituted wherever a bird is expected. Similarly, a union type is a supertype of each member of the set.

Changes required:

1. Allow if-else branches to return different types. The inferred type of the if-else expression will be a union of the two types. This can be represented using a new "Union" struct that can be defined [here](https://github.com/fika-lang/fika/blob/master/lib/fika/types.ex) like the FunctionRef struct. The to_string function of the struct will join the types separated by " | ".
2. When [inferring the return type of a function](https://github.com/fika-lang/fika/blob/master/lib/fika/type_checker.ex#L52), we'll now expand union types and add all possible function types in [add_function_type](https://github.com/fika-lang/fika/blob/master/lib/fika/type_checker.ex#L402). Example:

```elixir
# A function that takes x which can be either :foo or :bar and returns either :ok or :error
fn foo(x: :foo | :bar) : :ok | :error do
  # ...
end
```

The type checker will add the following function types to its "function_types" map in env:
1. module.foo(:foo) => :ok 
2. module.foo(:foo) => :error
3. module.foo(:bar) => :ok
4. module.foo(:bar) => :error 

We can optimize the storing and checking of function types later, but this should do for now.

Please let me know if I've missed something here!

### PR DESCRIPTION
**Pull request by polvalente**
New PR because the old one got too outdated with the new structs

closes #24

### CODE REVIEW

File: lib/fika/parser/non_literal_exps.ex [hunk 1]
```diff
@@ -49,6 +49,7 @@ defmodule Fika.Parser.NonLiteralExps do
     ])
     |> label("function call")
 
+  # TO-DO: accept nested if-else expressions
```
Comment (polvalente) [inline lib/fika/parser/non_literal_exps.ex:52]: @emilsoman During tests, I noticed that if-else doesn't allow an if-else expression inside itself. I'd rather deal with this in a separate PR

Comment (emilsoman) [reply lib/fika/parser/non_literal_exps.ex:52]: Sure!

File: lib/fika/type_checker.ex [hunk 2]
```diff
@@ -301,19 +301,13 @@ defmodule Fika.TypeChecker do
...
-
-      error =
-        "Expected if and else blocks to have same return type. " <>
-          "Got #{if_type_val} and #{else_type_val}"
-
-      {:error, error}
-    else
-      type
+    with {:ok, if_type_val, env} <- infer_block(env, if_block),
+         {:ok, else_type_val, _} <- infer_block(env, else_block) do
```
Comment (emilsoman) [inline lib/fika/type_checker.ex]: We need to carry forward all "env"s that we get as a result of type inference. In this line we are ignoring it using _.

File: test/fika/type_checker_test.exs [hunk 3]
```diff
@@ -508,6 +507,50 @@ defmodule Fika.TypeCheckerTest do
       assert {:ok, :Bool, _} = TypeChecker.infer(function, env)
     end
 
+    test "when function accepts union types and has if-else clause" do
```
Comment (emilsoman) [inline test/fika/type_checker_test.exs]: What do we mean when we say "function accepts union types"? If we are talking about the expected return type, we can say "function is expected to return union types"

File: test/fika/parser_test.exs [hunk 4]
```diff
@@ -501,6 +501,49 @@ defmodule Fika.ParserTest do
...
+                   }},
+                  [
+                    {{:if, {6, 114, 119}}, {:boolean, {2, 63, 72}, true},
+                     [{:atom, {3, 76, 83}, :ok}],
+                     [
+                       {:tuple, {5, 91, 113},
+                        [{:atom, {5, 91, 102}, :error}, {:string, {5, 91, 112}, ["not ok"]}]}
+                     ]}
+                  ]}}
+              ], "\n", %{}, {7, 120}, 123} = TestParser.function_def(code)
```
Comment (emilsoman) [inline test/fika/parser_test.exs]: We can use the new `function_def!` function to return the function def AST node directly.

**Review by emilsoman — COMMENTED** [review]
Apart from adding a test for flatten_types, the PR looks good to me! Looping in @fabionebbia to check out the parser changes because he recently worked on refactoring the parser.

File: lib/fika/type_checker.ex (hunk 2 above)
Comment (polvalente) [reply lib/fika/type_checker.ex]: nice, I'll fix this

**Pull request merged**